doc/config.rst
Configuration
=============

Configuration is handled with :code:`ade.Config` and can be modified for full
customization of the calculations. By default, high-level optimisations are
performed at PBE0-D3BJ/def2-SVP and single points at PBE0-D3BJ/def2-TZVP.


Calculations
------------

General
*******

The high-level electronic structure code defaults to the first available
from {ORCA, Gaussian09, Gaussian16, NWChem, QChem} and the low-level from
{XTB, MOPAC}. To select Gaussian09 as the high-level method:

.. code-block:: python

  >>> import autode as ade
  >>> ade.Config.hcode = 'g09'

Similarly, with the low-level:

.. code-block:: python

  >>> ade.Config.lcode = 'MOPAC'


To set the number of cores available and the memory per core (in MB), to use a maximum
of 32 GB for the whole calculation:

.. code-block:: python

  >>> ade.Config.n_cores = 8
  >>> ade.Config.max_core = 4000

------------

Keywords
********

**autodE** uses wrappers around common keywords used in QM calculations to allow
easy setting of e.g. a DFT functional.


.. code-block:: python

  >>> kwds = ade.Config.ORCA.keywords.sp
  >>> kwds.functional
  Functional(pbe0)


To modify the functional for single point energies, in ORCA:


.. code-block:: python

  >>> kwds.functional = 'B3LYP'


Alternatively, reassign to a whole new set of keywords:


.. code-block:: python

  >>> ade.Config.ORCA.keywords.sp = ['SP', 'B3LYP', 'def2-TZVP']


To add diffuse functions with the *ma* scheme to the def2-SVP default
basis set for optimisations:

.. code-block:: python

  >>> ade.Config.ORCA.keywords.set_opt_basis_set('ma-def2-SVP')


.. note::
    `set_opt_basis_set` sets the basis set in keywords.grad, keywords.opt_ts
    keywords.opt, keywords.low_opt and keywords.hess.

------------

XTB as a hmethod
****************

To use XTB as the *hmethod* for minima and TS optimisations within Gaussian use the `xtb-gaussian <https://github.com/aspuru-guzik-group/xtb-gaussian>`_ wrapper
and some default options. Note that the string to call `xtb-gaussian` will need to be modified with the appropriate keywords for spin and solvent, e.g., "xtb-gaussian --alpb water".

.. code-block:: python

  >>> kwds = ade.Config.G16.keywords
  >>> kwds.sp = ["External='xtb-gaussian'", "IOp(3/5=30)"]
  >>> kwds.low_opt = ["External='xtb-gaussian'", "Opt(Loose, NoMicro)", "IOp(3/5=30)"]
  >>> kwds.opt = ["External='xtb-gaussian'", "Opt(NoMicro)", "IOp(3/5=30)"]
  >>> kwds.opt_ts = ["External='xtb-gaussian'", "Opt(TS, CalcFC, NoEigenTest, MaxCycles=100, MaxStep=10, NoTrustUpdate, NoMicro)", "IOp(3/5=30)"]
  >>> kwds.hess = ["External='xtb-gaussian'", "Freq", "Geom(Redundant)", "IOp(3/5=30)"]
  >>> kwds.grad = ["External='xtb-gaussian'", 'Force(NoStep)', "IOp(3/5=30)"]

To use XTB within ORCA copy the :code:`xtb` binary to the folder where the :code:`orca` binary is located and rename it :code:`otool_xtb`, then
set the keywords to use. For example

.. code-block:: python

  >>> kwds = ade.Config.ORCA.keywords
  >>> kwds.sp = ['SP', 'PBE0', 'def2-SVP']
  >>> kwds.opt = ['Opt', 'XTB2']
  >>> kwds.low_opt = ['Opt', 'XTB2']
  >>> kwds.hess = ['NumFreq', 'XTB2']
  >>> kwds.grad = ['EnGrad', 'XTB2']
  >>> kwds.opt_ts = ['OptTS', 'NumFreq', 'XTB2\n',
    '%geom\n'
    'NumHess true\n'
    'Calc_Hess true\n'
    'Recalc_Hess 30\n'
    'Trust -0.1\n'
    'MaxIter 150\n'
    'end']


------------

Other
*****

See the `config file <https://github.com/duartegroup/autodE/blob/master/autode/config.py>`_
to see all the options.

.. note::
    NWChem currently only supports solvents for DFT, other methods must not have
    a solvent.

------------

Logging
-------

To set the logging level to one of {DEBUG, INFO, WARNING, ERROR} set the :code:`AUTODE_LOG_LEVEL`
environment variable, in bash::

    $ export AUTODE_LOG_LEVEL=INFO

To output the log to a file set e.g. *autode.log*::

    $ export AUTODE_LOG_FILE=autode.log

To log with timestamps and colours::

    $ conda install coloredlogs


To set the logging level permanently add the above export statements to
your *bash_profile*.


_sources/config.rst.txt
Configuration
=============

Configuration is handled with :code:`ade.Config` and can be modified for full
customization of the calculations. By default, high-level optimisations are
performed at PBE0-D3BJ/def2-SVP and single points at PBE0-D3BJ/def2-TZVP.


Calculations
------------

General
*******

The high-level electronic structure code defaults to the first available
from {ORCA, Gaussian09, Gaussian16, NWChem, QChem} and the low-level from
{XTB, MOPAC}. To select Gaussian09 as the high-level method:

.. code-block:: python

  >>> import autode as ade
  >>> ade.Config.hcode = 'g09'

Similarly, with the low-level:

.. code-block:: python

  >>> ade.Config.lcode = 'MOPAC'


To set the number of cores available and the memory per core (in MB), to use a maximum
of 32 GB for the whole calculation:

.. code-block:: python

  >>> ade.Config.n_cores = 8
  >>> ade.Config.max_core = 4000

------------

Keywords
********

**autodE** uses wrappers around common keywords used in QM calculations to allow
easy setting of e.g. a DFT functional.


.. code-block:: python

  >>> kwds = ade.Config.ORCA.keywords.sp
  >>> kwds.functional
  Functional(pbe0)


To modify the functional for single point energies, in ORCA:


.. code-block:: python

  >>> kwds.functional = 'B3LYP'


Alternatively, reassign to a whole new set of keywords:


.. code-block:: python

  >>> ade.Config.ORCA.keywords.sp = ['SP', 'B3LYP', 'def2-TZVP']


To add diffuse functions with the *ma* scheme to the def2-SVP default
basis set for optimisations:

.. code-block:: python

  >>> ade.Config.ORCA.keywords.set_opt_basis_set('ma-def2-SVP')


.. note::
    `set_opt_basis_set` sets the basis set in keywords.grad, keywords.opt_ts
    keywords.opt, keywords.low_opt and keywords.hess.

------------

XTB as a hmethod
****************

To use XTB as the *hmethod* for minima and TS optimisations with the `xtb-gaussian <https://github.com/aspuru-guzik-group/xtb-gaussian>`_ wrapper
and some default options. Note that the string to call `xtb-gaussian` will need to be modified with the appropriate keywords for spin and solvent, e.g., "xtb-gaussian --alpb water".

.. code-block:: python

  >>> ade.Config.G16.keywords.sp = ["External='xtb-gaussian'", "IOp(3/5=30)"]
  >>> ade.Config.G16.keywords.low_opt = ["External='xtb-gaussian'", "Opt(Loose, NoMicro)", "IOp(3/5=30)"]
  >>> ade.Config.G16.keywords.opt = ["External='xtb-gaussian'", "Opt(NoMicro)", "IOp(3/5=30)"]
  >>> ade.Config.G16.keywords.opt_ts = ["External='xtb-gaussian'", "Opt(TS, CalcFC, NoEigenTest, MaxCycles=100, MaxStep=10, NoTrustUpdate, NoMicro)", "IOp(3/5=30)"]
  >>> ade.Config.G16.keywords.hess = ["External='xtb-gaussian'", "Freq", "Geom(Redundant)", "IOp(3/5=30)"]
  >>> ade.Config.G16.keywords.grad = ["External='xtb-gaussian'", 'Force(NoStep)', "IOp(3/5=30)"]

------------

Other
*****

See the `config file <https://github.com/duartegroup/autodE/blob/master/autode/config.py>`_
to see all the options.

.. note::
    NWChem currently only supports solvents for DFT, other methods must not have
    a solvent.

------------

Logging
-------

To set the logging level to one of {INFO, WARNING, ERROR} set the AUTODE_LOG_LEVEL
environment variable, in bash::

    $ export AUTODE_LOG_LEVEL=INFO

To output the log to a file set e.g. *autode.log*::

    $ export AUTODE_LOG_FILE=autode.log

To log with timestamps and colours::

    $ conda install coloredlogs


To set the logging level permanently add the above export statements to
your *bash_profile*.


doc/changelog.rst
Changelog
=========

1.4.0
--------
----------

Functionality improvements
**************************
- Adds :code:`temporary_config()` context manager for temporary configuration changes


Usability improvements/Changes
******************************
- Adds full usability of autodE on Windows, including parallelisation with :code:`loky`
- Optional timeout for graph isomorphism test in Windows, turned on by :code:`Config.use_experimental_timeout=True` (default behaviour kept for Linux/macOS)
- The electronic temperature and the version of parameterisation for xTB calculations are made configurable


Bug Fixes
*********
- Fixes pickling issue with :code:`autode.config.Config` on Windows and in multiprocessing spawn for Linux/macOS


1.3.5
--------
----------


Usability improvements/Changes
******************************
- :code:`autode.value.ValueArray.to()` now defaults to copying the object rather than inplace modification


Functionality improvements
**************************
- Adds a :code:`to_` method to :code:`autode.value.ValueArray` for explicit inplace modification of the array


Bug Fixes
*********
- Fixes :code:`ERROR` logging level being ignored from environment variable :code:`AUTODE_LOG_LEVEL`
- Fixes :code:`autode.values.Value` instances generating items with units on division, and throw a warning if multiplying
- Fixes the printing of cartesian constraints in XTB input files, meaning they are no longer ignored
- Fixes :code:`Hessian` instances changing units when normal modes are calculated
- Fixes an incorrect alias for :code:`ev_per_ang`


1.3.4
--------
----------

Feature additions.

Usability improvements/Changes
******************************
* Throw useful exception for invalid :code:`ade.Config.ts_template_folder_path`


Functionality improvements
**************************
- Adds :code:`ade.transition_states.TransitionState.from_species` method to construct transition states from a species or molecule
- Adds :code:`autode.Reaction.save()` and :code:`autode.Reaction.load()` to save and reload a reaction state
- Adds saving checkpoints of a reaction during :code:`autode.Reaction.calculate_reaction_profile`


Bug Fixes
*********
- Fixes calculation :code:`clean_up()` failing with a null filename


1.3.3
--------
----------

Bugfix release.


Functionality improvements
**************************
- Adds skipping Hessian re-evaluation when using autodE optimisers if a molecules has a Hessian calculated at the same level of theory
- Adds a Hessian recalculation frequency to :code:`autode.optimisers.PRFOptimiser`
- Improves the default step size for TS optimising to be consistent with the ORCA default

Bug Fixes
*********
- Adds checking of SMILES-defined charge against the user-specified value
- Fixes :code:`autode.optimisers.CRFOptimiser` building incomplete internal coordinates for partially or completely fragmented molecules


1.3.2
--------
----------

Bugfix release.


Usability improvements/Changes
******************************
* Removes :code:`autode.geom.get_distance_constraints` in favour of a better named method :code:`TSBase().active_bond_constraints`


Bug Fixes
*********
- :code:`autode.transition_states.ts_guess.TSguess.from_species` now inherits solvent from the species
- Fixes a possible race condition in I/O for XTB conformer optimisations


1.3.1
--------
----------

Bugfix release.


Bug Fixes
*********
- Fixes behaviour of :code:`autode.utils.work_in_tmp_dir` and :code:`autode.utils.work_in` decorators
- Fixes an exception being raised when :code:`autode.Calculation.clean_up` is called with a method that doesn't implement external I/O
- Fixes autodE driven optimisations skipping execution when the input but not name changes


1.3.0
--------
----------

Optimisation features, graph assignment improvements and bugfixes.


Usability improvements/Changes
******************************
* Defines dummy atoms to have zero covalent and vdW radii
* Renames :code:`Method().available` to :code:`Method().is_available`
* Removes :code:`autode.bonds.get_ideal_bond_length_matrix` and :code:`autode.bonds.get_avg_bond_length`
* Removes :code:`autode.geom.rotate_columns`
* Modifies the names of most optimiser classes e.g. :code:`autode.opt.optimisers.PRFOOptimiser` -> :code:`PRFOptimiser`
* Simplifies initialising a :code:`autode.calculations.Calculation` by:

  * Requiring constraints to be attributed to a molecule
  * Removing the :code:`bond_ids_to_add` argument and using the labeled graph instead (active edges)
  * Removing the :code:`other_input_block` argument and appending to the keywords instead

* Removes :code:`autode.calculations.Calculation.print_final_output_lines` in favour of a method on :code:`calculation.output`
* Makes many methods in :code:`autode.calculations.Calculation` private
* Deprecates all :code:`autode.calculations.Calculation.get_<method>` methods in favour of setting properties of the input molecule
* Returns :code:`None` rather than rasies exceptions when calling the (deprecated) calculation get methods, to be consistent with ...get_energy
* Adds an :code:`autode.wrappers.keywords` package to improve file structure
* Removes any exceptions on calling :code:`.run()` on an optimiser instance where the system has no degrees of freedom
* Removes support for Python < v3.8
* Tweaks the default ORCA TS optimisation keywords to be more conservative, i.e. slower and more accurate


Functionality improvements
**************************
- Adds a :code:`autode.atoms.Atom.covalent_radius` property
- Adds a :code:`autode.atoms.Atoms.eqm_bond_distance` method for the equilibrium bonded distance between two atoms

autode/config.py
import os
from typing import Any
from autode.values import Frequency, Distance, Allocation
from autode.wrappers.keywords import implicit_solvent_types as solv
from autode.wrappers.keywords import KeywordsSet, MaxOptCycles
from autode.wrappers.keywords.basis_sets import (
    def2svp,
    def2tzvp,
    def2ecp,
    def2tzecp,
)
from autode.wrappers.keywords.functionals import pbe0
from autode.wrappers.keywords.dispersion import d3bj
from autode.wrappers.keywords.ri import rijcosx

location = os.path.abspath(__file__)


class _ConfigClass:
    # -------------------------------------------------------------------------
    # Total number of cores available
    #
    n_cores = 4
    # -------------------------------------------------------------------------
    # Per core memory available
    #
    max_core = Allocation(4, units="GB")
    # -------------------------------------------------------------------------
    # DFT code to use. If set to None then the highest priority available code
    # will be used:
    # 1. 'orca', 2. 'g09' 3. 'nwchem'
    #
    hcode = None
    # -------------------------------------------------------------------------
    # Semi-empirical/tight binding method to use. If set to None then the
    # highest priority available will be used:   1. 'xtb', 2. 'mopac'
    #
    lcode = None
    # -------------------------------------------------------------------------
    # When using explicit solvent is stable this will be uncommented
    #
    # explicit_solvent = False
    #
    # -------------------------------------------------------------------------
    # Setting to keep input files, otherwise they will be removed
    #
    keep_input_files = True
    # -------------------------------------------------------------------------
    # Use a different base directory for calculations with low-level methods
    # e.g. /dev/shm with a low level method, if None then will use the default
    # in tempfile.mkdtemp
    #
    ll_tmp_dir = None
    # -------------------------------------------------------------------------
    # By default templates are saved to /path/to/autode/transition_states/lib/
    # unless ts_template_folder_path is set
    #
    ts_template_folder_path = None
    # -------------------------------------------------------------------------
    # Whether or not to create and save transition state templates
    #
    make_ts_template = True
    # -------------------------------------------------------------------------
    # Save plots with dpi = 400
    #
    high_quality_plots = True
    # -------------------------------------------------------------------------
    # RMSD in angstroms threshold for conformers. Larger values will remove
    # more conformers that need to be calculated but also reduces the chance
    # that the lowest energy conformer is found
    #
    rmsd_threshold = Distance(0.3, units="Å")
    # -------------------------------------------------------------------------
    # Total number of conformers generated in find_lowest_energy_conformer()
    # for single molecules/TSs
    #
    num_conformers = 300
    # -------------------------------------------------------------------------
    # Maximum random displacement in angstroms for conformational searching
    #
    max_atom_displacement = Distance(4.0, units="Å")
    # -------------------------------------------------------------------------
    # Number of evenly spaced points on a sphere that will be used to generate
    # NCI and Reactant and Product complex conformers. Total number of
    # conformers will be:
    #   (num_complex_sphere_points ×
    #              num_complex_random_rotations) ^ (n molecules in complex - 1)
    #
    num_complex_sphere_points = 10
    # -------------------------------------------------------------------------
    # Number of random rotations of a molecule that is added to a NCI or
    # Reactant/Product complex
    #
    num_complex_random_rotations = 10
    # -------------------------------------------------------------------------
    # For more than 2 molecules in a complex the conformational space explodes,
    # so limit the maximum number to this value
    #
    max_num_complex_conformers = 300
    # -------------------------------------------------------------------------
    # Use the high + low level method to find the lowest energy
    # conformer, to use energies at the low_opt level of the low level code
    # set this to False
    #
    hmethod_conformers = True
    # -------------------------------------------------------------------------
    # Set to True to use single point energy evaluations to rank conformers and
    # select the lowest energy. Requires keywords.low_sp to be set and
    # hmethod_conformers = True
    # WARNING: This relies on the low-level geometry being accurate enough for
    # the system in question – switching this on without benchmarking may lead
    # to large errors!
    #
    hmethod_sp_conformers = False
    # -------------------------------------------------------------------------
    # Use adaptive force constant modification in NEB calculations to improve
    # sampling around the saddle point
    #
    adaptive_neb_k = True
    # -------------------------------------------------------------------------
    # Minimum and maximum step size to use for the adaptive path search
    #
    min_step_size = Distance(0.05, units="Å")
    max_step_size = Distance(0.3, units="Å")
    # -------------------------------------------------------------------------
    # Heuristic for pruning the bond rearrangement set. If there are only bond
    # rearrangements that involve small rings then TSs involving small rings
    # are possible. However, when there are multiple possibilities involving
    # the same set of atoms then discard any rearrangements that would involve
    # a 3 or 4-membered TS e.g. skip the possible 4-membered TS for a Cope
    # rearrangement in hexadiene
    #
    skip_small_ring_tss = True
    # -------------------------------------------------------------------------
    # Minimum magnitude of the imaginary frequency (cm-1) to consider for a
    # 'true' TS. For very shallow saddle points this may need to be reduced
    # to e.g. -10 cm-1. Although most TSs have |v_imag| > 100 cm-1 this
    # threshold is designed to be conservative
    #
    min_imag_freq = Frequency(-40, units="cm-1")
    # -------------------------------------------------------------------------
    # Configuration parameters for ideal gas free energy calculations. Can be
    # configured to use different standard states, quasi-rigid rotor harmonic
    # oscillator (qRRHO) or pure RRHO
    #
    #  One of: '1M', '1atm'
    standard_state = "1M"
    #
    # Method to treat low frequency modes (LFMs). Either standard RRHO ('igm'),
    # Truhlar's method where all frequencies below a threshold are scaled to
    # a shifted value (see J. Phys. Chem. B, 2011, 115, 14556) or Grimme's
    # method of interpolating between HO and RR (i.e. qRRHO, see
    # Chem. Eur. J. 2012, 18, 9955)
    #
    # One of: 'igm', 'truhlar', 'grimme'
    lfm_method = "grimme"
    #
    # Parameters for Grimme's method (only used when lfm_method='grimme'),
    # w0 is a frequency in cm-1
    grimme_w0 = Frequency(100, units="cm-1")
    grimme_alpha = 4
    #
    # Parameters for Truhlar's method (only used when lfm_method='truhlar')
    # vibrational frequencies below this value (cm-1) will be shifted to this
    # value before the entropy is calculated
    vib_freq_shift = Frequency(100, units="cm-1")
    # -------------------------------------------------------------------------
    # Frequency scale factor, useful for DFT functions known to have a
    # systematic error. This value must be between 0 and 1 inclusive. For
    # example, PBEh-3c has a scale factor of 0.95.
    #
    freq_scale_factor = None
    # -------------------------------------------------------------------------
    # Minimum number of atoms that are removed for truncation to be used in
    # locating TSs. Below this number any truncation is skipped
    #
    min_num_atom_removed_in_truncation = 10
    # -------------------------------------------------------------------------
    # Flag for allowing free energies to be calculated with association
    # complexes. This is *not* recommended to be turned on due to the
    # approximations made in the entropy calculations.
    #
    allow_association_complex_G = False
    # -------------------------------------------------------------------------
    # Flag to allow use of an experimental timeout function wrapper for
    # Windows, using loky. The default case has no timeout for Windows, and
    # timeout only works on Linux/macOS. This flag is ignored on Linux/macOS.
    #
    use_experimental_timeout = False
    # -------------------------------------------------------------------------

    class ORCA:
        # ---------------------------------------------------------------------
        # Parameters for orca   https://sites.google.com/site/orcainputlibrary/
        # ---------------------------------------------------------------------
        #
        # Path can be unset and will be assigned if it can be found in $PATH
        path = None

        optts_block = (
            "\n%geom\n"
            "Calc_Hess true\n"
            "Recalc_Hess 20\n"
            "Trust -0.1\n"
            "MaxIter 100\n"
            "end"
        )

        keywords = KeywordsSet(
            low_opt=[
                "LooseOpt",
                pbe0,
                rijcosx,
                d3bj,
                def2svp,
                "def2/J",
                MaxOptCycles(10),
            ],
            grad=["EnGrad", pbe0, rijcosx, d3bj, def2svp, "def2/J"],
            low_sp=["SP", pbe0, rijcosx, d3bj, def2svp, "def2/J"],
            opt=["Opt", pbe0, rijcosx, d3bj, def2svp, "def2/J"],
            opt_ts=[
                "OptTS",
                "Freq",
                pbe0,
                rijcosx,
                d3bj,
                def2svp,
                "def2/J",
                optts_block,
            ],
            hess=["Freq", pbe0, rijcosx, d3bj, def2svp, "def2/J"],
            sp=["SP", pbe0, rijcosx, d3bj, def2tzvp, "def2/J"],
            ecp=def2ecp,
        )

        # Implicit solvent in ORCA is either treated with CPCM or SMD, the
        # former has support for a VdW surface construction which provides
        # better geometry convergence (https://doi.org/10.1002/jcc.26139) SMD
        # is in general more accurate, but does not (yet) have support for the
        # VdW charge scheme. Use either (1) solv.cpcm, (2) solv.smd
        implicit_solvation_type = solv.cpcm

    class G09:
        # ---------------------------------------------------------------------
        # Parameters for g09                 https://gaussian.com/glossary/g09/
        # ---------------------------------------------------------------------
        #
        # path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        grid = "integral=ultrafinegrid"
        optts_block = (
            "Opt=(TS, CalcFC, NoEigenTest, MaxCycles=100, "
            "MaxStep=10, NoTrustUpdate)"
        )

        keywords = KeywordsSet(
            low_opt=[pbe0, def2svp, "Opt=Loose", MaxOptCycles(10), d3bj, grid],
            grad=[pbe0, def2svp, "Force(NoStep)", d3bj, grid],
            low_sp=[pbe0, def2svp, d3bj, grid],
            opt=[pbe0, def2svp, "Opt", d3bj, grid],
            opt_ts=[pbe0, def2svp, "Freq", d3bj, grid, optts_block],
            hess=[pbe0, def2svp, "Freq", d3bj, grid],
            sp=[pbe0, def2tzvp, d3bj, grid],
            ecp=def2tzecp,
        )

        # Only SMD implemented
        implicit_solvation_type = solv.smd

    class G16:
        # ---------------------------------------------------------------------
        # Parameters for g16                   https://gaussian.com/gaussian16/
        # ---------------------------------------------------------------------
        #
        # path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        ts_str = (
            "Opt=(TS, CalcFC, NoEigenTest, MaxCycles=100, MaxStep=10, "
            "NoTrustUpdate, RecalcFC=30)"
        )

        keywords = KeywordsSet(
            low_opt=[pbe0, def2svp, "Opt=Loose", d3bj, MaxOptCycles(10)],
            grad=[pbe0, def2svp, "Force(NoStep)", d3bj],
            low_sp=[pbe0, def2svp, d3bj],
            opt=[pbe0, def2svp, "Opt", d3bj],
            opt_ts=[pbe0, def2svp, "Freq", d3bj, ts_str],
            hess=[pbe0, def2svp, "Freq", d3bj],
            sp=[pbe0, def2tzvp, d3bj],
            ecp=def2tzecp,
        )

        # Only SMD implemented
        implicit_solvation_type = solv.smd

    class NWChem:
        # ---------------------------------------------------------------------
        # Parameters for nwchem    http://www.nwchem-sw.org/index.php/Main_Page
        # ---------------------------------------------------------------------
        #
        # Path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        # Note that the default NWChem level is PBE0 and PBE rather than
        # PBE0-D3BJ and PBE-D3BJ as only D3 is available. The optimisation
        # keywords contain 'gradient' as the optimisation is driven by autodE
        keywords = KeywordsSet(
            low_opt=[def2svp, pbe0, MaxOptCycles(10), "task dft gradient"],
            grad=[def2svp, pbe0, "task dft gradient"],
            low_sp=[def2svp, pbe0, "task dft energy"],
            opt=[def2svp, pbe0, MaxOptCycles(100), "task dft gradient"],
            opt_ts=[def2svp, pbe0, MaxOptCycles(50), "task dft gradient"],
            hess=[def2svp, pbe0, "task dft freq"],
            sp=[def2tzvp, pbe0, "task dft energy"],
            ecp=def2ecp,
        )

        # Only SMD implemented
        implicit_solvation_type = solv.smd

    class XTB:
        # ---------------------------------------------------------------------
        # Parameters for xtb                  https://github.com/grimme-lab/xtb
        # ---------------------------------------------------------------------
        #
        # path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        keywords = KeywordsSet()
        #
        # Only GBSA implemented
        implicit_solvation_type = solv.gbsa
        #
        # Force constant used for harmonic restraints in constrained
        # optimisations (Ha/a0)
        force_constant = 2
        #
        # Electronic temperature for all calculations (Kelvin)
        # None means unset (default), set to 300.0 to have 300K for example
        etemp = None
        #
        # Version of xTB hamiltonian parameterisation: 0,1 or 2
        # corresponding to GFN0-xTB, GFN1-xTB, GFN2-xTB respectively
        # When unset, uses the default
        gfn_ver = None

    class MOPAC:
        # ---------------------------------------------------------------------
        # Parameters for mopac                             http://openmopac.net
        # ---------------------------------------------------------------------
        #
        # path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        # Note: all optimisations at this low level will be in the gas phase
        # using the keywords_list specified here. Solvent in mopac is defined
        # by EPS and the dielectric
        keywords = KeywordsSet(low_opt=["PM7", "PRECISE"])
        #
        # Only COSMO implemented
        implicit_solvation_type = solv.cosmo

    class QChem:
        # ---------------------------------------------------------------------
        # Parameters for QChem                          https://www.q-chem.com/
        # ---------------------------------------------------------------------
        #
        # path can be unset and will be assigned if it can be found in $PATH
        path = None
        #
        # Default set of keywords to use for different types of calculation
        keywords = KeywordsSet(
            low_opt=[pbe0, def2svp, "jobtype opt", MaxOptCycles(10), d3bj],
            grad=[pbe0, def2svp, "jobtype force", d3bj],
            low_sp=[pbe0, def2svp, d3bj],
            opt=[pbe0, def2svp, "jobtype opt", d3bj],
            opt_ts=[pbe0, def2svp, "jobtype TS", d3bj],
            hess=[pbe0, def2svp, "jobtype Freq", d3bj],
            sp=[pbe0, def2tzvp, d3bj],
            ecp=def2ecp,
        )

        #
        # Only SMD is implemented
        implicit_solvation_type = solv.smd

    # =========================================================================
    # =============               End                        ==================
    # =========================================================================

    def __setattr__(self, key, value):
        """Custom setters"""

        if not hasattr(self, key):
            raise KeyError(f"Cannot set {key}. Not present in ade.Config")

        if key == "max_core":
            value = Allocation(value).to("MB")

        if key == "freq_scale_factor":

            if value is not None:
                if not (0.0 < value <= 1.0):
                    raise ValueError(
                        "Cannot set the frequency scale factor "
                        "outside of (0, 1]"
                    )

                value = float(value)

        if key in ("max_atom_displacement", "min_step_size", "max_step_size"):
            if float(value) < 0:
                raise ValueError(f"Distances cannot be negative. Had: {value}")

            value = Distance(value).to("ang")

        return super().__setattr__(key, value)


def _instantiate_config_opts(cls: type) -> Any:
    """
    Instantiate a config class containing options defined
    as class variables. It generates an instance of the
    class, and then creates instance variables of the same
    name as class variables, recursively converting any
    nested class into instances.
    (This is required because class variables are not pickled,
    only instance variables are)

    Args:
        cls (type): Must be a class containing class
         variables (not instance)

    Returns:
        (Any): The generated class instance
    """
    if not isinstance(cls, type):
        raise ValueError("Must be a class, not an instance")
    cls_instance = cls()
    for name, attr in cls.__dict__.items():
        if name.startswith("__"):
            continue
        if isinstance(attr, type):
            attr_val = _instantiate_config_opts(attr)  # recursive
        else:
            attr_val = attr
        setattr(cls_instance, name, attr_val)
    return cls_instance


# Single instance of the configuration
Config = _instantiate_config_opts(_ConfigClass)
